leftwm-core/src/handlers/command_handler/scratchpad_handler.rs

```diff
@@ -0,0 +1,1211 @@
+            if direction == Direction::Forward {
+                scratchpad_windows.push_front(window);
+            } else {
+                scratchpad_windows.push_back(window);
+            }
+            return Some(window);
+        }
+
+        log::info!(
+            "Dead window in scratchpad found, discard: window PID: {}",
+            window
+        );
+    }
+
+    None
+}
+
+/// Check if the scratchpad is visible on the current tag.
+/// If the name of the scratchpad is incorrect, false will be returned
+fn is_scratchpad_visible<C: Config, SERVER: DisplayServer>(
+    manager: &Manager<C, SERVER>,
+    scratchpad_name: &str,
+) -> bool {
+    // Like Try operator but returns false and only works on `Option`s
+    macro_rules! try_bool {
+        ($cond:expr) => {
+            if let Some(value) = $cond {
+                value
+            } else {
+                return false;
+            }
+        };
+    }
+    let current_tag = try_bool!(manager.state.focus_manager.tag(0));
+    let scratchpad = try_bool!(manager.state.active_scratchpads.get(scratchpad_name));
+
+    // Filter out all the non existing windows (invalid pid) and map to window
+    // Check if any of them is in the current tag
+    scratchpad
+        .iter()
+        .filter_map(|pid| manager.state.windows.iter().find(|w| w.pid == Some(*pid)))
+        .any(|window| window.has_tag(&current_tag))
+}
+
+/// Handle the command to toggle the scratchpad
+pub fn toggle_scratchpad<C: Config, SERVER: DisplayServer>(
+    manager: &mut Manager<C, SERVER>,
+    name: &str,
+) -> Option<bool> {
+    let current_tag = &manager.state.focus_manager.tag(0)?;
+    let scratchpad = manager
+        .state
+        .scratchpads
+        .iter()
+        .find(|s| name == s.name)?
+        .clone();
+
+    if let Some(id) = manager.state.active_scratchpads.get_mut(&scratchpad.name) {
+        if let Some(first_in_scratchpad) =
+            next_valid_scratchpad_pid(id, &manager.state.windows, Direction::Forward)
+        {
+            if let Some((is_visible, window_handle)) = manager
+                .state
+                .windows
+                .iter()
+                .find(|w| w.pid == Some(first_in_scratchpad))
+                .map(|w| (w.has_tag(current_tag), w.handle))
+            {
+                if is_visible {
+                    // Window is visible => Hide the scratchpad.
+                    if let Err(msg) = hide_scratchpad(manager, &window_handle) {
+                        log::error!("{}", msg);
+                        return Some(false);
+                    }
+                } else {
+                    // Window is hidden => show the scratchpad
+                    if let Err(msg) = show_scratchpad(manager, &window_handle) {
+                        log::error!("{}", msg);
+                        return Some(false);
+                    }
+                }
+
+                return Some(true);
```

Comment:
I think this can be summarized as:
```rust
    if let Some((is_visible, window_handle)) = manager.state
        .windows
        .iter()
        .find(|w| w.pid == Some(first_in_scratchpad))
        .map(|w| (w.has_tag(current_tag), w.handle))
    {
        let action_result = if is_visible {
            hide_scratchpad(manager, &window_handle)
        } else {
            show_scratchpad(manager, &window_handle)
        };

        match action_result {
            Ok(()) => Some(true),
            Err(msg) => {
                log::error!("{}", msg);
                Some(false)
            },
        }
```